File: 1448. Count Good Nodes in Binary Tree/other2.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @description https://leetcode.com/problems/count-good-nodes-in-binary-tree/
 * @tutorial https://leetcode.com/problems/count-good-nodes-in-binary-tree/submissions/
 * @param {TreeNode} root
 * @return {number}
 */
var goodNodes = function(root) {
  var count = 0;
  dfs(root, -Infinity);

  function dfs(node, maxSoFar) {
    if (maxSoFar <= node.val) {
      count++;
      maxSoFar = node.val;
    }
    if (node.left) {
      dfs(node.left, maxSoFar);
    }
    if (node.right) {
      dfs(node.right,maxSoFar);
    }
  }

  return count;
}

/**
 * try to enhance the perf of other.js
 */

File: 1448. Count Good Nodes in Binary Tree/other2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
/**
 * https://leetcode.com/problems/count-good-nodes-in-binary-tree/
 * https://leetcode.com/problems/count-good-nodes-in-binary-tree/solution/
 */
class Pair {
  public TreeNode node;
  public int maxSoFar;
  
  public Pair(TreeNode node, int maxSoFar) {
    this.node = node;
    this.maxSoFar = maxSoFar;
  }
}

class Solution {
  public int goodNodes(TreeNode root) {
    int numGoodNodes = 0;
    Queue<Pair> queue = new LinkedList<>();
    queue.offer(new Pair(root, Integer.MIN_VALUE));
    
    while (queue.size() > 0) {
      Pair curr = queue.poll();
      if (curr.maxSoFar <= curr.node.val) {
        numGoodNodes++;
      }
      
      if (curr.node.right != null) {
        queue.offer(new Pair(curr.node.right, Math.max(curr.node.val, curr.maxSoFar)));
      }

      if (curr.node.left != null) {
        queue.offer(new Pair(curr.node.left, Math.max(curr.node.val, curr.maxSoFar)));
      }
    }

    return numGoodNodes;
  }
}